### DISCUSSION TITLE
Configuration Management for Z-wave device configuration and associations

### DISCUSSION THREAD

**Discussion by LordMike** [zwave-js/zwave-js-ui#2137] [Show and tell]
Hey all,

When I want to do something with z-wave, or any system, I prototype it using code, such that I can quickly iterate and work on finding out how stuff works and how best to do it. Some of these things move on to feature requests, if I think those features could benefit others.

I wanted to show off three of these features that I made.

Note: For those not in the know: "Configuration Management" is a discipline where configuration of devices, services and systems is centrally managed and rolled out. The manager will know which configurations have "drifted" (changed outside the system), and thereby reapply them as needed. The benefit of config management is streamlining, as in you'll centrally be able to configure "all" of something.

Note2: The code and tool is _not_ available right now. I _could_ make it available.

## Configuration management
I've made a tool that I can run, which uses the MQTT API of Zwavejs2mqtt to query all nodes (`getNodes` api) and thereby know all devices and their settings. I then have a config file like the following:

<details>
  <summary>Example devices config</summary>

```
### ZDB5100
# Enable dimming
# 0 No dimming
# 1 Trailing edge dimming. (Default)
# 2 Leading edge dimming
product:ZDB5100 112-0-4	1
name:wallswitch_30_2	112-0-4	0 # Disable dimmer on some switches

# Remove default colors
product:ZDB5100	112-0-22-255	0
product:ZDB5100	112-0-22-65280	0
product:ZDB5100	112-0-22-16711680	0
product:ZDB5100	112-0-22-4278190080	0
product:ZDB5100	112-0-23-255	0
product:ZDB5100	112-0-23-65280	0
product:ZDB5100	112-0-23-16711680	0
product:ZDB5100	112-0-23-4278190080	0
product:ZDB5100	112-0-30-255	0
product:ZDB5100	112-0-30-65280	0
product:ZDB5100	112-0-30-16711680	0
product:ZDB5100	112-0-30-4278190080	0
product:ZDB5100	112-0-31-255	0
product:ZDB5100	112-0-31-65280	0
product:ZDB5100	112-0-31-16711680	0
product:ZDB5100	112-0-31-4278190080	0
product:ZDB5100	112-0-38-255	0
product:ZDB5100	112-0-38-65280	0
product:ZDB5100	112-0-38-16711680	0
product:ZDB5100	112-0-38-4278190080	0
product:ZDB5100	112-0-39-255	0
product:ZDB5100	112-0-39-65280	0
product:ZDB5100	112-0-39-16711680	0
product:ZDB5100	112-0-39-4278190080	0
product:ZDB5100	112-0-46-255	0
product:ZDB5100	112-0-46-65280	0
product:ZDB5100	112-0-46-16711680	0
product:ZDB5100	112-0-46-4278190080	0
product:ZDB5100	112-0-47-255	0
product:ZDB5100	112-0-47-65280	0
product:ZDB5100	112-0-47-16711680	0
product:ZDB5100	112-0-47-4278190080	0

product:ZDB5100	112-0-10	0 # Disable enhanced LED control
product:ZDB5100	112-0-5	10  # Minimum dimming level
product:ZDB5100	112-0-2	2 # Dimming over Ns
product:ZDB5100	112-0-3	0 # On/off takes Ns

# Button 1 controls light
product:ZDB5100	112-0-1-1	1
product:ZDB5100	112-0-1-2	0
product:ZDB5100	112-0-1-4	0
product:ZDB5100	112-0-1-8	0

# Button 2 controls light
name:wallswitch_1	112-0-1-1	0
name:wallswitch_1	112-0-1-2	1
name:wallswitch_22	112-0-1-1	0
name:wallswitch_22	112-0-1-2	1

### Sensative Strips
product:11-01-011	112-0-1	0 # Send Binary Sensor reports
product:11-01-011	112-0-2	0 # Disable LED

### Aeotec door window sensor 7
# http://manual.zwave.eu/backend/make.php?lang=EN&sku=AEOEZWA008&cert=
# http://manuals-backend.z-wave.info/make.php?lang=en&sku=ZWA008-C&cert=ZC10-19066568
product:ZWA008	112-0-14	0 # Disable tilt sensor

### Aeotec doorbell
# http://manuals-backend.z-wave.info/make.php?lang=en&sku=ZW162-C&cert=ZC10-19046416&type=mini
# Sound levels
product:ZW162	121-1-defaultVolume	0
product:ZW162	121-2-defaultVolume	0
product:ZW162	121-3-defaultVolume	25  # Doorbell 1
product:ZW162	121-4-defaultVolume	0
product:ZW162	121-5-defaultVolume	0
product:ZW162	121-6-defaultVolume	0
product:ZW162	121-7-defaultVolume	0
product:ZW162	121-8-defaultVolume	0

....
```
</details>

This config snippet configures a _lot_ of settings on my ZDB5100's (wall switches with LEDs), and for some devices, it sets up different settings than the "default" (the lines with `product:ZDB5100`). Each configuration line "appends" to the previous, so you can make generic configuration first, and then more specific configuration later.

In my case, I want to do a lot to a new ZDB5100 - I want to disable the default LED's that light up, and then change stuff like the dimming duration (5 seconds by default) or which button controls the current dimmer (usually 1, top-left - but sometimes I want it differently)..

Using this, I can run a quick command on my system to run my code, which then queries all values and changes those as needed:

> zwave-mqtt-templater device-configs -d .\DeviceConfigsFile.txt

## Associations management
I haven't quite figured out a good way to model associations in general, but this is currently for me more of a way to "save" associations and "restore" them. This can help when I want to test out things, or re-include all my devices using S2 (which factory resets everything)..

Like before, I can run commands:

<details>
  <summary>Example associations commands</summary>

```
# Save all current associations, add `--refresh` to requery all associations
zwave-mqtt-templater associations-get -f Associations.json

# Save all current associations for only some devices
# This updates only the selected devices in the file, so it can be used to update the file with new stuff
zwave-mqtt-templater associations-get -f Associations.json product:ZDB5100

# Restore associations from file
zwave-mqtt-templater associations-set -f Associations.json

# Restore associations from file, only for some devices
zwave-mqtt-templater associations-set -f Associations.json product:ZDB5100
```
</details>

The json file looks like this:

<details>
  <summary>Example associations config</summary>

```jsonc
{
  "Nodes": {
    // Each node uses its name
    "wallswitch_1": {
      "Links": {
         // Endpoint 0, group 1 has the following partners
        "0.1": [
           // If the partner has no name, we just put the node id
           // This relationship is for node 1, endpoint 0
          "1.0"
        ],
        "1.4": [
           // For partners where we can determine the node name, it is used in place of an id
          "wallswitch_22.5"
        ],
        "5.2": [
          "wallswitch_22.4"
        ]
      }
    },
    "wallswitch_30_2": {
      "Links": {
        "0.1": [
          "1.0"
        ],
        "1.4": [
          "wallswitch_30_1.5"
        ]
      }
    },
    "sensor_50_1": {
      "Links": {
        "0.1": [
           // This relationship is for node 1, but with no endpoint
          "1.-"
        ]
      }
    },
    // If the node has no name, the node id is used
    "81": {
      "Links": {
        "0.1": [
          "1.-"
        ]
      }
    }
  }
}
```
</details>

## Configuration overview
I was troubleshooting an issue where one device was acting differently from the others. When I looked into this, I couldn't spot _why_ it was acting differently, so I turned to configurations again.. something must be off. I made a `dump-config` that again uses `getNodes` and looks at a set of devices (the same model f.ex.), and writes out all configurations and their current values. I can then plot this in Excel, and actually get a nice table of each setting and device (since all devices have the same settings), and quickly identify the odd on out.

Using that, I found out that at some point I had tweaked a setting I didn't manage, manually on one device - the one acting out. It meant that its LED acted differently than other devices, so I couldn't control it like I normally do.

<details>
  <summary>Dump config output</summary>

The actual data
![image](https://user-images.githubusercontent.com/1027111/147959428-e510eb08-863a-4bbe-821e-0b72f5287725.png)

My pivot table in excel that groups by setting
![image](https://user-images.githubusercontent.com/1027111/147959350-3381ccc1-4278-4133-b631-9f89b121e0f4.png)
</details>

## Benefits
This may seem like a lot of work, but it has benefits when messing with or tweaking the system. For example, recently, I wanted to reinclude some devices with S2 to try it out. This resets the devices, but I can quickly restore both configurations and associations, so that my home continues working as intended. All I need to do, is maintain the naming of my devices, and then I can reapply.

**Comment by robertsLando** [comment]
@LordMike Thanks! BTW cannot find the links to the tool :)

**Reply by LordMike** [reply]
I didn't release it .. I should probably have been more explicit about that..

I _could_ make it available though, would it be interesting? :)

**Reply by robertsLando** [reply]
It may be useful for sure :)

**Reply by LordMike** [reply]
_You_, on the other hand, should consider if these things are feature-worthy for the web UI ... :)

**Reply by LordMike** [reply]
I've made the tool have an actual commandline and made releases (https://github.com/LordMike/MBW.Tools.ZwaveJs2Mqtt).

I think it works on Linux, but I haven't tested :P
